Log week 13 (#65)
elias-sundqvist/DATX02-20-04: docs/log.org

  The time log has exact tasks worked on each entry, for each
* Week 13
We have made a lot of progress this week on all three components, that being
1. Audio generation
2. Structure generation
3. Combination of above
** Audio generation
To start, here are some of the same note generated from piano notes over the
course of 513K steps.

#+begin_columns
#+begin_column
*44K*

{{{AUDIO(audio/44k.wav)}}}
#+end_column
#+begin_column
*62K*

{{{AUDIO(audio/62k.wav)}}}
#+end_column
#+begin_column
*83K*

{{{AUDIO(audio/83k.wav)}}}
#+end_column
#+end_columns

#+begin_columns
#+begin_column
*153K*

{{{AUDIO(audio/153k.wav)}}}
#+end_column
#+begin_column
*230K*

{{{AUDIO(audio/230k.wav)}}}
#+end_column
#+begin_column
*513K*

{{{AUDIO(audio/513k.wav)}}}
#+end_column
#+end_columns

They sound pretty good, but still very artificial. These are from spectrograms
of size $128 \times 256$.

*** Pitch conditioning
Due to our audio results not being good enough, we experimented with a deeper
model with more convolution layers. This ended up not being that easy to train,
and had trouble learning anything at all.

The shallow model used last week had a lot of trouble with classifying pitches.

> The primary reason we think is the resolution begin too smaller. To demonstrate

Reviewer: ```suggestion
The primary reason we think is the resolution being too low. To demonstrate
```

> this, we have increased the resolution of spectrograms to $128 \times 256$, it

Reviewer: ```suggestion
this, we have increased the resolution of spectrograms to $128 \times 256$, which
```

> double the number of frequency 'bins'.

Reviewer: ```suggestion
doubles the number of frequency 'bins'.
```

#+CAPTION:A mel spectrogram of every pitch available in the gansynth dataset. These are generated pitches from training our shallow model, so they are not perfect but they demonstrate how resolution matters. The label number indicates MIDI pitch. Note how a higher pitch decreases space between overtones.
#+NAME: fig:high_res_pitch
[[./img/high_res_pitch.png]]

The spectrograms in figure [[fig:high_res_pitch]] are generated using our shallow
model and the higher resolution. As we have already explained, the horizontal
lines are overtones. Between a tone and an overtone, there is a finite amount of
space that has to be large enough for 11 other notes, those being the other
musical notes (12 tones in total).A larger resolution ensures these 11 notes fit
into the spectrogram bins, which is vital for the auxiliary classifier to
recognise them.

The second, somewhat larger issue is that the space between overtones change
depending on the pitch of note. The nth overtone frequency is calculated as
$$f_{n} = n \times f_{0}$$ where $f_{0}$ is the base frequency of the note (or
pitch). As such, a very low pitch has very little space between the overtones

> than a high note. Even though the spectrogram uses logarithmic scale on the

Reviewer: ```suggestion
compared to a high note. Even though the spectrogram uses logarithmic scale on the
```

#+name: fig:512_example
[[./img/512.png]]

*** Constant Q transform
As shown before, we have been having issues with some pitches not generating
particularly well. The theory is that the resolution is not high enough for the
model to differentiate between two pitches. To solve this problem, we have to
delve deep into signal processing. Apologies if the following section is not
very explanatory, we have not had a course in this.

Right now, we use a simple fast fourier transform to generate our mel
spectrograms. This is fast and easy to do, but keeps resolution constant across
all frequencies. We visualised this last week, and there stated that we had
trouble particularly with the lower frequencies having enough space. The
~Constant Q transform~ somewhat solves this by letting the window length be a
multiple of the previous window. The window then grows with each bin, making it
larger for higher frequencies and smaller for lower. This means the
transform works better for our hearing, which follows a log scale.

> To use this, we simple switch which function to use. Figure [[fig:cqt_exp_1]]

Reviewer: ```suggestion
To use this, we simply switch which function to use. Figure [[fig:cqt_exp_1]]
```

2. Mediocre results

The following is not an original idea and first used by gansynth. One way to
solve both these issues is not having to rebuild the phase. Instead, we let the
network generate it. This sounds more difficult than it actually is. All we have
to do is double the channels models into (magnitude, phase), and ensure the
dataset is processed to not throw away the phase. Recreating the transformed
signal is as easy as $\text{mag} \times \text{phase}$ as you would using
phasors.

The more difficult problem is how to get the network to correctly learn phase.
It is easy for the network to learn magnitude because it is simply mapping the
color of a pixel in the image and activating it whenever it learnt to. Phase is
not constant for audio, the same pitch and note can have different phases
depending on the actual audio recording. This means just feeding the network
pitch would force it to map an infinite number of phases to our pitches, which
is not very easy for it to learn. This is where instantaneous frequency comes
in, which we have maybe mentioned before in this log. Basically, just take the
derivative of the phase, which remains constant. Then, the second channel

> describes earlier is just another image channel for the GAN to learn, no more

Reviewer: ```suggestion
described earlier is just another image channel for the GAN to learn, no more
```
